doc/new_architecture.md

```diff
@@ -0,0 +1,161 @@
+these changes are recorded in [a discussion in Agama's repository][drop-cockpit].
+
+[drop-cockpit]: https://github.com/openSUSE/agama/discussions/1000
+
+But before describing how the architecture should look, let's quickly look at the current status.
+
+## The current architecture
+
+At this point, Agama is composed of three high-level components:
+
+* **Agama service**: implements the logic to perform the installation. It is the core component of
+Agama and it offers a D-Bus interface. Actually, it is composed of two services: `rubygem-agama` and
+`agama-dbus-server`.
+
+* **Web user interface (`cockpit-agama`)**: a web-based interface that plays the role of a GUI when
+using Agama live.
+
+* **Command Line Interface (`agama-cli`)**: it allows to interact with Agama core and drives the
+auto-installation process.
+
+In addition to those components, we need to consider Cockpit, which plays a vital role:
+
+* It makes communication between the web UI and the D-Bus services possible.
+* It makes the web UI code available to the browser.
+* It takes care of authenticating the user when connecting remotely. Again, it is only relative to
+the web UI.
+
+```mermaid
+flowchart LR
+  subgraph Clients
+    Browser
+    CLI
+  end
+
+  subgraph Cockpit
+    UI["Web UI"]
+  end
+
+  subgraph Agama Service
+    Rust[Agama Rust]
+    Ruby[Agama Ruby]
+  end
+
+  Browser <---> UI
+  UI <--D-Bus--> Rust
+  UI <--D-Bus--> Ruby
+  CLI <--D-Bus-->  Rust
+  CLI <--D-Bus--> Ruby
+  Rust <--D-Bus---> Ruby
+```
+
+## The new architecture
+
+The proposed architecture is not that different from the current one, but it tries to meet these
+goals:
+
+* Drop our dependency on Cockpit.
+* Implement a higher-level API to be consumed by the clients, partially replacing the D-Bus.
+
+### Components
+
+With those goals in mind, we are considering the following components:
+
+* **Agama core (old `agama-dbus-server`)**: implements the installation logic. It relies heavily on
+the Agama YaST service.
+
+* **Agama YaST service (old `rubygem-agama`)**: it is written in Ruby and has direct access to YaST
+libraries. Complex parts, like storage and software handling, are implemented in this component.
+
+* **HTTP and WebSocket API**: implements the API the clients should use to communicate with Agama.
+
+* **Web user interface (old `cockpit-agama`)**: Agama's graphical user interface. The web server
+makes this React application available to the browsers.
+
+* **Command Line Interface (`agama-cli`)**: it allows interaction with Agama core and drives the
+auto-installation process. With the new architecture, connecting through the network might be
+possible without SSH.
+
+The following diagram could be better, but it represents the main components and their interactions.
+
+```mermaid
+flowchart LR
+  subgraph Clients
+    Browser
+    CLI
+  end
+
+  subgraph Agama Core
+    subgraph Web["Web Server"]
+      direction TB
+      UI["Web UI"]
```

Comment:
it is kind of strange to have in one layer UI, API and WS. Also I am a bit confused why websocket and API is used by browser and not by UI? I kind of expect that browser communicate with UI and UI ( aka js code ) do that calls to API and WS.

Reply:
Well, it depends where you put the bar :-) Actually, the browser itself is handling the HTTP/WS stuff. You are serving the UI (a static HTML/JavaScript/CSS file set), the HTTP API and the WebSocket.